src/Dialect/ONNX/Transforms/FusePadAvgpool.cpp

```diff
@@ -0,0 +1,147 @@
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+#include "src/Pass/Passes.hpp"
+
+using namespace mlir;
+using namespace onnx_mlir;
+
+namespace {
+
+struct FusePadIntoAveragePoolPattern
+    : public OpRewritePattern<ONNXAveragePoolOp> {
+  using OpRewritePattern<ONNXAveragePoolOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXAveragePoolOp avgOp, PatternRewriter &rewriter) const override {
+
+    Value input = avgOp.getX();
+    auto padOp = input.getDefiningOp<ONNXPadOp>();
+    if (!padOp)
+      return failure();
+
+    // Check that pad mode is "constant" (default value, so should never be
+    // null)
+    StringRef mode = padOp.getMode();
+    if (mode != "constant")
+      return failure();
+    float padValue = 0.0f;
+
+    Value padsInput = padOp.getPads();
+    Value constantValInput = padOp.getConstantValue();
+
+    auto padsConstOp =
+        dyn_cast_or_null<ONNXConstantOp>(padsInput.getDefiningOp());
+    if (!padsConstOp)
+      return failure();
+    auto padsAttr = dyn_cast_or_null<ElementsAttr>(padsConstOp.getValueAttr());
+    if (!padsAttr)
+      return failure();
+
+    auto constOp =
+        dyn_cast_or_null<ONNXConstantOp>(constantValInput.getDefiningOp());
+    if (!constOp)
+      return failure();
+    auto constAttr = dyn_cast_or_null<ElementsAttr>(constOp.getValueAttr());
+
+    if (!constAttr)
+      return failure();
+
+    auto firstAttr = *constAttr.getValues<Attribute>().begin();
+    if (auto fAttr = mlir::dyn_cast<FloatAttr>(firstAttr))
+      padValue = fAttr.getValueAsDouble();
+
+    if (padValue != 0.0f)
+      return failure();
+
+    // Only handle 4D tensors (NCHW format)
+    auto inputType = padOp.getData().getType().dyn_cast<RankedTensorType>();
+    if (!inputType || inputType.getRank() != 4)
+      return failure();
```

Comment:
Please do not use the .cast member methods. They are deprecated and will be removed in the future
```suggestion
    // Only handle 4D tensors 
    auto inputType = dyn_cast<RankedTensorType>(padOp.getData().getType());
    if (!inputType || inputType.getRank() != 4)
      return failure();
```

Reply:
done